inet: get resolv to handle multiple AAAA records

 redo mainstream fixes in gethostbyname_r() to gethostbyname2_r()

 patch from TomatoUSB project by Wes Campaigne with minor modifications.

---
 libc/inet/resolv.c |   88 +++++++++++++++++++++++++++++++++++++----------------
 1 file changed, 62 insertions(+), 26 deletions(-)

--- uClibc-0.9.30.1/libc/inet/resolv.c
+++ uClibc/libc/inet/resolv.c
@@ -2106,8 +2106,10 @@ int gethostbyname_r(const char * name,
 		__set_errno(0);         /* to check for missing /etc/hosts. */
 		i = __get_hosts_byname_r(name, AF_INET, result_buf,
 				buf, buflen, result, h_errnop);
-		if (i == 0)
+		if (i == NETDB_SUCCESS) {
+			__set_errno(old_errno);
 			return i;
+		}
 		switch (*h_errnop) {
 			case HOST_NOT_FOUND:
 			case NO_ADDRESS:
@@ -2260,10 +2262,10 @@ int gethostbyname2_r(const char *name, i
 #else /* __UCLIBC_HAS_IPV6__ */
 	struct in6_addr *in;
 	struct in6_addr **addr_list;
+	char **alias;
 	unsigned char *packet;
 	struct resolv_answer a;
 	int i;
-	int nest = 0;
 	int __nameserversXX;
 	char ** __nameserverXX;
 	int wrong_af = 0;
@@ -2274,7 +2276,6 @@ int gethostbyname2_r(const char *name, i
 	if (family != AF_INET6)
 		return EINVAL;
 
-	__open_nameservers();
 	*result = NULL;
 	if (!name)
 		return EINVAL;
@@ -2287,7 +2287,7 @@ int gethostbyname2_r(const char *name, i
 		i = __get_hosts_byname_r(name, family, result_buf,
 				buf, buflen, result, h_errnop);
 		if (i == NETDB_SUCCESS) {
-//FIXME: restore errno?
+			__set_errno(old_errno);
 			return i;
 		}
 		switch (*h_errnop) {
@@ -2308,6 +2308,15 @@ int gethostbyname2_r(const char *name, i
 
 	DPRINTF("Nothing found in /etc/hosts\n");
 
+	/* make sure user char * is aligned */
+	i = ALIGN_BUFFER_OFFSET(buf);
+	if (unlikely(i)) {
+		if (buflen < i)
+			return ERANGE;
+		buf += i;
+		buflen -= i;
+	}
+
 	*h_errnop = NETDB_INTERNAL;
 	if (buflen < sizeof(*in))
 		return ERANGE;
@@ -2324,16 +2333,26 @@ int gethostbyname2_r(const char *name, i
 	addr_list[0] = in;
 	addr_list[1] = 0;
 
+	if (buflen < sizeof(char *)*ALIAS_DIM)
+		return ERANGE;
+	alias = (char **)buf;
+	buf += sizeof(char **)*ALIAS_DIM;
+	buflen -= sizeof(char **)*ALIAS_DIM;
+
 	if (buflen < 256)
 		return ERANGE;
 	strncpy(buf, name, buflen);
 
+	alias[0] = buf;
+	alias[1] = NULL;
+
 	/* First check if this is already an address */
 	if (inet_pton(AF_INET6, name, in)) {
 		result_buf->h_name = buf;
 		result_buf->h_addrtype = AF_INET6;
 		result_buf->h_length = sizeof(*in);
 		result_buf->h_addr_list = (char **) addr_list;
+		result_buf->h_aliases = alias;
 		*result = result_buf;
 		*h_errnop = NETDB_SUCCESS;
 		return NETDB_SUCCESS;
@@ -2347,56 +2366,72 @@ int gethostbyname2_r(const char *name, i
 		return TRY_AGAIN;
 	}
 
-	memset(&a, '\0', sizeof(a));
+	__open_nameservers();
 
-	for (;;) {
+        /*for (;;)*/ {
+//FIXME: why was it a loop? It never loops...
 		__UCLIBC_MUTEX_LOCK(__resolv_lock);
 		__nameserversXX = __nameservers;
 		__nameserverXX = __nameserver;
 		__UCLIBC_MUTEX_UNLOCK(__resolv_lock);
-
-		i = __dns_lookup(buf, T_AAAA, __nameserversXX, __nameserverXX, &packet, &a);
+		a.buf = buf;
+		a.buflen = buflen;
+		a.add_count = 0;
+		i = __dns_lookup(name, T_AAAA, __nameserversXX, __nameserverXX, &packet, &a);
 
 		if (i < 0) {
 			*h_errnop = HOST_NOT_FOUND;
+			DPRINTF("__dns_lookup\n");
 			return TRY_AGAIN;
 		}
 
-		strncpy(buf, a.dotted, buflen);
-		free(a.dotted);
-
-		if (a.atype == T_CNAME) {		/* CNAME */
-			DPRINTF("Got a CNAME in gethostbyname()\n");
-			i = __decode_dotted(packet, a.rdoffset, buf, buflen);
+		if ((a.rdlength + sizeof(struct in6_addr*)) * a.add_count + 256 > buflen) {
+			free(a.dotted);
 			free(packet);
+			*h_errnop = NETDB_INTERNAL;
+			DPRINTF("buffer too small for all addresses\n");
+			return ERANGE;
+		}
 
-			if (i < 0) {
-				*h_errnop = NO_RECOVERY;
-				return -1;
-			}
-			if (++nest > MAX_RECURSE) {
-				*h_errnop = NO_RECOVERY;
-				return -1;
-			}
-			continue;
+		if (a.add_count > 0) {
+			memmove(buf - sizeof(struct in6_addr*)*2, buf, a.add_count * a.rdlength);
+			addr_list = (struct in6_addr**)(buf + a.add_count * a.rdlength);
+			addr_list[0] = in;
+			for (i = a.add_count - 1; i >= 0; --i)
+				addr_list[i+1] = (struct in6_addr*)(buf - sizeof(struct in6_addr*)*2 + a.rdlength * i);
+			addr_list[a.add_count + 1] = 0;
+			buflen -= (((char*)&(addr_list[a.add_count + 2])) - buf);
+			buf = (char*)&addr_list[a.add_count + 2];
 		}
-		if (a.atype == T_AAAA) {	/* ADDRESS */
+
+		strncpy(buf, a.dotted, buflen);
+		free(a.dotted);
+
+		if (a.atype == T_AAAA) { /* ADDRESS */
 			memcpy(in, a.rdata, sizeof(*in));
 			result_buf->h_name = buf;
 			result_buf->h_addrtype = AF_INET6;
 			result_buf->h_length = sizeof(*in);
 			result_buf->h_addr_list = (char **) addr_list;
+#ifdef __UCLIBC_MJN3_ONLY__
+#warning TODO -- generate the full list
+#endif
+			result_buf->h_aliases = alias; /* TODO: generate the full list */
 			free(packet);
-			break;
+			/*was: break;*/
+			*result = result_buf;
+			*h_errnop = NETDB_SUCCESS;
+			return NETDB_SUCCESS;
 		}
 		free(packet);
 		*h_errnop = HOST_NOT_FOUND;
 		return TRY_AGAIN;
 	}
-
+/*
 	*result = result_buf;
 	*h_errnop = NETDB_SUCCESS;
 	return NETDB_SUCCESS;
+*/
 #endif /* __UCLIBC_HAS_IPV6__ */
 }
 libc_hidden_def(gethostbyname2_r)
-- 
